contracts/TokenDistributor.sol

```diff
@@ -0,0 +1,158 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {StakingRewardsV2} from "./StakingRewardsV2.sol";
+import {RewardEscrowV2} from "./RewardEscrowV2.sol";
+import {IKwenta} from "./interfaces/IKwenta.sol";
+
+contract TokenDistributor {
+    /// @notice event for tracking new epochs
+    event NewEpochCreated(uint block, uint epoch);
+
+    /// @notice error when user tries to create a new distribution too soon
+    error CannotCreateNewDistributionThisSoon();
+
+    /// @notice tracks block, previously claimed fees,
+    /// kwenta balance of this contract, and total
+    /// staked amount in StakingRewardsV2
+    struct Distribution {
+        uint epochStartBlockNumber;
+        uint epochStartTime;
+        uint previouslyClaimedFees;
+        uint kwentaStartOfEpoch;
+    }
+
+    /// @notice tracks the distribution for each epoch
+    mapping(uint => Distribution) public distributionEpochs;
+
+    /// @notice represents the status of if a person already
+    /// claimed their epoch
+    mapping(address => mapping(uint => bool)) public claimedEpochs;
+
+    /// @notice kwenta interface
+    IKwenta public kwenta;
+
+    /// @notice Counter for new epochs (starts at 0)
+    uint256 public epoch;
+
+    /// @notice running total for claimed fees
+    uint256 public claimedFees;
+
+    /// @notice rewards staking contract
+    StakingRewardsV2 public stakingRewardsV2;
+
+    /// @notice escrow contract which holds (and may stake) reward tokens
+    RewardEscrowV2 public rewardEscrowV2;
+
+    constructor(
+        address _kwenta,
+        address _stakingRewardsV2,
+        address _rewardEscrowV2
+    ) {
+        kwenta = IKwenta(_kwenta);
+        epoch = 0;
+        stakingRewardsV2 = StakingRewardsV2(_stakingRewardsV2);
+        rewardEscrowV2 = RewardEscrowV2(_rewardEscrowV2);
+    }
+
+    /// @notice  creates a new Distribution entry at the current block,
+    /// can only be called once per week
+    function newDistribution() public {
+        ///@dev [epoch - 1] to get the start of last weeks epoch
+        if (epoch > 0) {
+            require(
+                block.timestamp >=
+                    (distributionEpochs[epoch - 1].epochStartTime + 1 weeks),
+                "TokenDistributor: Last week's epoch has not ended yet"
+            );
+        }
+
+        Distribution memory distribution = Distribution(
+            block.number,
+            block.timestamp,
+            claimedFees,
+            kwenta.balanceOf(address(this))
+        );
+        //for a given block it might not be the final value of the block
+        distributionEpochs[epoch] = distribution;
+        //todo: timestamp is put in for block but that is technically the time
+        emit NewEpochCreated(block.timestamp, epoch);
+
+        epoch++;
```

Comment:
you can `emit NewEpochCreated(block.timestamp, epoch++);` it's the same thing

Reply:
i emit before incrementing because the uint256 epoch actually leads ahead, so it always the next epoch to start or the one that will happen after the current one. but i've been thinking about changing this because it can be confusing and relates to the comments above and below.